pkgs/objective_c/src/input_stream_adapter.m

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+#import "input_stream_adapter.h"
+
+#import <Foundation/Foundation.h>
+#include <os/log.h>
+
+@implementation DartInputStreamAdapter {
+  Dart_Port _sendPort;
+  NSCondition *_dataCondition;
+  NSMutableData *_data;
+  NSStreamStatus _status;
+  BOOL _done;
+  NSError *_error;
+  id<NSStreamDelegate> _delegate; // This is a weak reference.
```

Comment:
Does this need a `__weak` annotation? I thought ARC would treat this as a strong ref by default.

Also, if we're dealing with weak ref stuff, the test probably needs a ref counting test to make sure it's all working as expected. There aren't any ref counting tests in package:objective_c yet, so you'll have to copy across the utils from ffigen (`objectRetainCount` and `doGC`). Take a look at pkgs/ffigen/test/native_objc_test/arc_test.dart to see how they work. Ignore the stuff that uses `counter` (that's a bit redundant since I added `objectRetainCount`). Just check that `objectRetainCount` > 0, drop your Dart references, do a GC, then check that `objectRetainCount` == 0.

Reply:
> Does this need a `__weak` annotation? I thought ARC would treat this as a strong ref by default.

By default, `_delegate` is `self` so I think that you need the reference to be weak to prevent the circular reference:

https://developer.apple.com/library/archive/documentation/General/Conceptual/DevPedia-CocoaCore/ObjectOwnership.html


> Also, if we're dealing with weak ref stuff, the test probably needs a ref counting test to make sure it's all working as expected. There aren't any ref counting tests in package:objective_c yet, so you'll have to copy across the utils from ffigen (`objectRetainCount` and `doGC`). Take a look at pkgs/ffigen/test/native_objc_test/arc_test.dart to see how they work. Ignore the stuff that uses `counter` (that's a bit redundant since I added `objectRetainCount`). Just check that `objectRetainCount` > 0, drop your Dart references, do a GC, then check that `objectRetainCount` == 0.

Reply:
> By default, `_delegate` is `self` so I think that you need the reference to be weak to prevent the circular reference:
> 
> https://developer.apple.com/library/archive/documentation/General/Conceptual/DevPedia-CocoaCore/ObjectOwnership.html

Yeah, I mean shouldn't you add an explicit `__weak` annotation? The way it's written now, I think ARC will be treating this field as strong.

Reply:
Got it. Fixed.